frontend/app/controllers/spree/checkout_controller.rb

```diff
@@ -25,33 +25,62 @@ class CheckoutController < Spree::StoreController
 
     # Updates the order and advances to the next state (when possible.)
     def update
-      if OrderUpdateAttributes.new(@order, update_params, request_env: request.headers.env).apply
-        @order.temporary_address = !params[:save_user_address]
-        success = if @order.state == 'confirm'
-          @order.complete
-        else
-          @order.next
-        end
-        if !success
-          flash[:error] = @order.errors.full_messages.join("\n")
-          redirect_to(checkout_state_path(@order.state)) && return
-        end
+      if update_order
+
+        assign_temp_address
+
+        redirect_on_failure and return unless transition_forward!
+
 
         if @order.completed?
-          @current_order = nil
-          flash.notice = Spree.t(:order_processed_successfully)
-          flash['order_completed'] = true
-          redirect_to completion_route
+          finalize_order
         else
-          redirect_to checkout_state_path(@order.state)
+          send_to_next_state
         end
+
       else
         render :edit
       end
     end
 
     private
 
+    def update_order
+      OrderUpdateAttributes.new(@order, update_params, request_env: request.headers.env).apply
+    end
+
+    def assign_temp_address
+      @order.temporary_address = !params[:save_user_address]
+    end
+
+    def redirect_on_failure
+      flash[:error] = @order.errors.full_messages.join("\n")
+      redirect_to(checkout_state_path(@order.state))
+    end
+
+    def transition_forward!
```

Comment:
I don't think this should be a "bang" methd. There's no non-bang version of this. We're calling `next` and `complete` on order, not `next!` and `complete!`.

I'd prefer it was just called `transition_forward` or similar.